src/interface/swarm.hpp

```diff
@@ -0,0 +1,150 @@
+#ifndef INTERFACE_SWARM_HPP_
+#define INTERFACE_SWARM_HPP_
+
+///
+/// A swarm contains all particles of a particular species
+/// Date: August 21, 2019
+
+#include "basic_types.hpp"
+#include "bvals/cc/bvals_cc.hpp"
+#include "metadata.hpp"
+#include "parthenon_arrays.hpp"
+#include "variable.hpp"
+#include <array>
+#include <cstdint>
+#include <list>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace parthenon {
+class MeshBlock;
+
+enum class PARTICLE_STATUS { UNALLOCATED, ALIVE, DEAD };
+
+class Swarm {
+ public:
+  MeshBlock *pmy_block = nullptr; // ptr to MeshBlock
+
+  Swarm(const std::string label, const Metadata &metadata, const int nmax_pool_in = 3);
+
+  // TODO BRR This should really be const... mask_ is managed internally
+  /// Get mask array for active particles
+  ParticleVariable<bool> &GetMask() { return mask_; }
+
+  /// Get the mask array of active particles that are marked for removal
+  ParticleVariable<bool> &GetMarkedForRemoval() { return marked_for_removal_; }
+
+  /// Make a new Swarm based on an existing one
+  std::shared_ptr<Swarm> AllocateCopy(const bool allocComms = false,
+                                      MeshBlock *pmb = nullptr);
+
+  /// Add variable to swarm
+  void Add(const std::string label, const Metadata &metadata);
+
+  /// Add multiple variables with common metadata to swarm
+  void Add(const std::vector<std::string> labelVector, const Metadata &metadata);
+
+  /// Remote a variable from swarm
+  void Remove(const std::string label);
+
+  /// Get real particle variable
+  ParticleVariable<Real> &GetReal(const std::string label) {
+    return *(realMap_.at(label));
+  }
+
+  /// Get integer particle variable
+  ParticleVariable<int> &GetInteger(const std::string label) {
+    return *(intMap_.at(label));
+  }
+
+  /// Assign label for swarm
+  void setLabel(const std::string label) { label_ = label; }
+
+  /// retrieve label for swarm
+  std::string label() const { return label_; }
+
+  /// retrieve metadata for swarm
+  const Metadata metadata() const { return m_; }
+
+  /// Assign info for swarm
+  void setInfo(const std::string info) { info_ = info; }
+
+  /// return information string
+  std::string info() { return info_; }
+
+  /// Expand pool size geometrically as necessary
+  void increasePoolMax() { setPoolMax(2 * nmax_pool_); }
+
+  /// Set max pool size
+  void setPoolMax(const int nmax_pool);
+
+  /// Check whether metadata bit is set
+  bool IsSet(const MetadataFlag bit) const { return m_.IsSet(bit); }
+
+  /// Get the last index of active particles
+  int get_max_active_index() { return max_active_index_; }
+
+  /// Get number of active particles
+  int get_num_active() { return num_active_; }
+
+  /// Get the quality of the data layout. 1 is perfectly organized, < 1
+  /// indicates gaps in the list.
+  Real get_packing_efficiency() { return num_active_ / (max_active_index_ + 1); }
+
+  bool mpiStatus;
+
```

Comment:
I would prefer to encapsulate this with methods but won't make an argument in this case.